src/link/assign.cpp

```diff
@@ -179,20 +178,43 @@ static struct FreeSpace *getPlacement(struct Section const *section,
 
 			// If that location is past the current block's end,
 			// go forwards until that is no longer the case.
-			while (space && location->address >=
-						space->address + space->size)
+			while (space && location->address >= space->address + space->size)
 				space = space->next;
 
 			// Try again with the new location/free space combo
 		}
 
-		if (section->isBankFixed)
+		// Try again in the next bank, if one is available.
+		// Try scrambled banks in descending order until no bank in the scrambled range is available.
+		// Otherwise, try in ascending order.
+		if (section->isBankFixed) {
 			return NULL;
-
-		// Try again in the next bank
-		location->bank++;
-		if (location->bank > sectionTypeInfo[section->type].lastBank)
+		} else if (scrambleROMX && section->type == SECTTYPE_ROMX && location->bank <= scrambleROMX) {
+			if (location->bank > sectionTypeInfo[section->type].firstBank)
+				location->bank--;
+			else if (scrambleROMX < sectionTypeInfo[section->type].lastBank)
+				location->bank = scrambleROMX + 1;
+			else
+				return NULL;
+		} else if (scrambleWRAMX && section->type == SECTTYPE_WRAMX && location->bank <= scrambleWRAMX) {
+			if (location->bank > sectionTypeInfo[section->type].firstBank)
+				location->bank--;
+			else if (scrambleWRAMX < sectionTypeInfo[section->type].lastBank)
+				location->bank = scrambleWRAMX + 1;
+			else
+				return NULL;
+		} else if (scrambleSRAM && section->type == SECTTYPE_SRAM && location->bank <= scrambleSRAM) {
+			if (location->bank > sectionTypeInfo[section->type].firstBank)
+				location->bank--;
+			else if (scrambleSRAM < sectionTypeInfo[section->type].lastBank)
+				location->bank = scrambleSRAM + 1;
+			else
+				return NULL;
+		} else if (location->bank < sectionTypeInfo[section->type].lastBank) {
+			location->bank++;
```

Comment:
I hate that all of this logic is largely duplicated, and also feels very "action at a distance" (the logic that requires *this* to switch directions is far away). But I can swallow that pill considering that this logic isn't in our backwards compat promise (so we can implement other strategies) and we're going to rewrite this soon anyway.

Reply:
I don't like it either, but didn't see a way to abstract out the repetition that wouldn't cost more in complexity than it saves.

Reply:
Not in C, anyway. (Macros could be involved, but let's not.)